File: test_a2a_client.ts
/**
 * A2A Protocol Client Test (TypeScript/JavaScript)
 *
 * Tests the A2A endpoints from a TypeScript/JavaScript client.
 * Can be run with ts-node or compiled to JavaScript.
 *
 * Usage:
 *   npx ts-node test_a2a_client.ts
 *   npx ts-node test_a2a_client.ts --api-key YOUR_KEY
 */

interface AgentCard {
  id: string
  name: string
  description: string
  capabilities: string[]
  protocol_version: string
  metadata?: {
    endpoints?: Record<string, string>
    authentication?: {
      required: boolean
      methods: string[]
    }
    [key: string]: any
  }
}

interface MessageRequest {
  content: string
  format?: 'plain' | 'markdown' | 'code'
  session_id?: string
  context?: Record<string, any>
}

interface MessageResponse {
  message_id: string
  content: string
  format: string
  timestamp: string
  metadata?: {
    session_id?: string
    [key: string]: any
  }
}

class A2AClient {
  private baseUrl: string
  private apiKey?: string
  private sessionId?: string

  constructor(baseUrl: string = 'http://localhost:8000', apiKey?: string) {
    this.baseUrl = baseUrl.replace(/\/$/, '')
    this.apiKey = apiKey
  }

  private getHeaders(): Record<string, string> {
    const headers: Record<string, string> = {
      'Content-Type': 'application/json',
    }

    if (this.apiKey) {
      headers['Authorization'] = `Bearer ${this.apiKey}`
    }

    return headers
  }

  /**
   * Test T099: Get agent card
   */
  async getAgentCard(): Promise<AgentCard | null> {
    console.log('\n📋 Test 1: Getting Agent Card...')
    console.log(`   URL: ${this.baseUrl}/api/v1/a2a/agent-card`)

    try {
      const response = await fetch(`${this.baseUrl}/api/v1/a2a/agent-card`)

      if (response.ok) {
        const agentCard: AgentCard = await response.json()
        console.log('   ✅ Success!')
        console.log(`   Agent ID: ${agentCard.id}`)
        console.log(`   Agent Name: ${agentCard.name}`)
        console.log(`   Capabilities: ${agentCard.capabilities.join(', ')}`)
        console.log(`   Protocol Version: ${agentCard.protocol_version}`)
        return agentCard
      } else {
        console.log(`   ❌ Failed: HTTP ${response.status}`)
        const text = await response.text()
        console.log(`   Response: ${text}`)
        return null
      }
    } catch (error) {
      console.log(`   ❌ Error: ${error instanceof Error ? error.message : String(error)}`)
      return null
    }
  }

  /**
   * Test T100: Send message to agent
   */
  async sendMessage(content: string, format: 'plain' | 'markdown' | 'code' = 'plain'): Promise<MessageResponse | null> {
    console.log('\n💬 Test 2: Sending Message...')
    console.log(`   URL: ${this.baseUrl}/api/v1/a2a/message`)
    console.log(`   Content: "${content}"`)
    console.log(`   Format: ${format}`)

    try {
      const payload: MessageRequest = {
        content,
        format,
      }

      if (this.sessionId) {
        payload.session_id = this.sessionId
      }

      const response = await fetch(`${this.baseUrl}/api/v1/a2a/message`, {
        method: 'POST',
        headers: this.getHeaders(),
        body: JSON.stringify(payload),
      })

      if (response.ok) {
        const messageResponse: MessageResponse = await response.json()
        console.log('   ✅ Success!')
        console.log(`   Message ID: ${messageResponse.message_id}`)
        console.log(`   Response Format: ${messageResponse.format}`)
        console.log(`   Response Content:`)
        console.log(`   ${messageResponse.content.substring(0, 200)}...`)

        // Extract session_id
        if (messageResponse.metadata?.session_id) {
          this.sessionId = messageResponse.metadata.session_id
          console.log(`   Session ID: ${this.sessionId}`)
        }

        return messageResponse
      } else if (response.status === 401) {
        console.log('   ❌ Authentication Failed (401)')
        const json = await response.json()
        console.log(`   Response: ${JSON.stringify(json)}`)
        console.log('   💡 Hint: You may need to provide an API key')
        return null
      } else {
        console.log(`   ❌ Failed: HTTP ${response.status}`)
        const text = await response.text()
        console.log(`   Response: ${text}`)
        return null
      }
    } catch (error) {
      console.log(`   ❌ Error: ${error instanceof Error ? error.message : String(error)}`)
      return null
    }
  }

  /**
   * Test T101: Validate agent card format compliance
   */
  validateAgentCardCompliance(agentCard: AgentCard): boolean {
    console.log('\n🔍 Test 3: Validating Agent Card Compliance...')

    const requiredFields = ['id', 'name', 'description', 'capabilities', 'protocol_version']
    const missingFields = requiredFields.filter((field) => !(field in agentCard))

    if (missingFields.length > 0) {
      console.log(`   ❌ Missing required fields: ${missingFields.join(', ')}`)
      return false
    }

    // Check capabilities
    if (!Array.isArray(agentCard.capabilities)) {
      console.log("   ❌ 'capabilities' must be an array")
      return false
    }

    // Check metadata
    if (agentCard.metadata && !agentCard.metadata.endpoints) {
      console.log("   ⚠️  Warning: 'metadata.endpoints' not found")
    }

    console.log('   ✅ Agent card format is compliant!')
    return true
  }

  /**
   * Health check
   */
  async checkHealth(): Promise<boolean> {
    console.log('\n🏥 Health Check...')
    console.log(`   URL: ${this.baseUrl}/api/v1/a2a/health`)

    try {
      const response = await fetch(`${this.baseUrl}/api/v1/a2a/health`)

      if (response.ok) {
        const health = await response.json()
        console.log('   ✅ Service is healthy')
        console.log(`   Status: ${health.status}`)
        console.log(`   Protocol: ${health.protocol}`)
        console.log(`   Version: ${health.version}`)
        console.log(`   Active Sessions: ${health.active_sessions}`)
        return true
      } else {
        console.log(`   ❌ Health check failed: HTTP ${response.status}`)
        return false
      }
    } catch (error) {
      console.log(`   ❌ Error: ${error instanceof Error ? error.message : String(error)}`)
      return false
    }
  }
}

async function runTests(baseUrl: string, apiKey?: string): Promise<boolean> {
  console.log('='.repeat(60))
  console.log('🚀 A2A Protocol Client Test Suite (TypeScript)')
  console.log('='.repeat(60))
  console.log(`Base URL: ${baseUrl}`)
  console.log(`Authentication: ${apiKey ? 'Enabled (API Key provided)' : 'Disabled'}`)
  console.log('='.repeat(60))

  const client = new A2AClient(baseUrl, apiKey)

  // Test 0: Health check
  if (!(await client.checkHealth())) {
    console.log('\n❌ Server is not available. Please start the server first.')
    console.log('   Run: cd backend && python -m src.server.app')
    return false
  }

  // Test 1: Get agent card
  const agentCard = await client.getAgentCard()
  if (!agentCard) {
    console.log('\n❌ Failed to get agent card')
    return false
  }

  // Test 3: Validate agent card
  if (!client.validateAgentCardCompliance(agentCard)) {
    return false
  }

  // Test 2: Send messages
  const testMessages: Array<[string, 'plain' | 'markdown' | 'code']> = [
    ['Hello! How are you?', 'plain'],
    ['What can you do?', 'plain'],
    ['Show me a code example', 'plain'],
  ]

  for (const [content, format] of testMessages) {
    const response = await client.sendMessage(content, format)
    if (!response) {
      console.log(`\n❌ Failed to send message: ${content}`)
      return false
    }
  }

  console.log('\n' + '='.repeat(60))
  console.log('✅ All tests passed successfully!')
  console.log('='.repeat(60))
  console.log('\n📊 Summary:')
  console.log('   • Agent card retrieved and validated')
  console.log('   • Messages sent and received')
  console.log(`   • Session maintained: ${client['sessionId']}`)
  console.log('   • A2A protocol compliance verified')

  return true
}

// Main execution
async function main() {
  const args = process.argv.slice(2)
  const baseUrlIndex = args.indexOf('--base-url')
  const apiKeyIndex = args.indexOf('--api-key')

  const baseUrl = baseUrlIndex !== -1 && args[baseUrlIndex + 1] ? args[baseUrlIndex + 1] : 'http://localhost:8000'
  const apiKey = apiKeyIndex !== -1 && args[apiKeyIndex + 1] ? args[apiKeyIndex + 1] : undefined

  try {
    const success = await runTests(baseUrl, apiKey)
    process.exit(success ? 0 : 1)
  } catch (error) {
    console.log(`\n\n❌ Unexpected error: ${error instanceof Error ? error.message : String(error)}`)
    process.exit(1)
  }
}

// Run if executed directly
if (require.main === module) {
  main()
}

export { A2AClient, AgentCard, MessageRequest, MessageResponse }

File: backend/test_a2a_client.py
#!/usr/bin/env python3
"""
A2A Protocol Client Test Script

Tests the A2A endpoints from an external client perspective.
This simulates how another application would interact with our A2A agent.

Usage:
    python test_a2a_client.py                    # Test without authentication
    python test_a2a_client.py --api-key YOUR_KEY # Test with authentication
"""

import argparse
import requests
import json
from typing import Optional
import sys


class A2AClient:
    """
    Simple A2A protocol client for testing
    """

    def __init__(self, base_url: str = "http://localhost:8000", api_key: Optional[str] = None):
        self.base_url = base_url.rstrip('/')
        self.api_key = api_key
        self.session_id: Optional[str] = None

    def _get_headers(self) -> dict:
        """Get request headers with optional authentication"""
        headers = {"Content-Type": "application/json"}
        if self.api_key:
            headers["Authorization"] = f"Bearer {self.api_key}"
        return headers

    def get_agent_card(self) -> dict:
        """
        Test T099: Get agent card

        GET /api/v1/a2a/agent-card
        """
        print("\n📋 Test 1: Getting Agent Card...")
        print(f"   URL: {self.base_url}/api/v1/a2a/agent-card")

        try:
            response = requests.get(
                f"{self.base_url}/api/v1/a2a/agent-card",
                timeout=10
            )

            if response.status_code == 200:
                agent_card = response.json()
                print("   ✅ Success!")
                print(f"   Agent ID: {agent_card.get('id')}")
                print(f"   Agent Name: {agent_card.get('name')}")
                print(f"   Capabilities: {', '.join(agent_card.get('capabilities', []))}")
                print(f"   Protocol Version: {agent_card.get('protocol_version')}")
                return agent_card
            else:
                print(f"   ❌ Failed: HTTP {response.status_code}")
                print(f"   Response: {response.text}")
                return {}

        except Exception as e:
            print(f"   ❌ Error: {str(e)}")
            return {}

    def send_message(self, content: str, message_format: str = "plain") -> Optional[dict]:
        """
        Test T100: Send message to agent

        POST /api/v1/a2a/message
        """
        print(f"\n💬 Test 2: Sending Message...")
        print(f"   URL: {self.base_url}/api/v1/a2a/message")
        print(f"   Content: \"{content}\"")
        print(f"   Format: {message_format}")

        try:
            payload = {
                "content": content,
                "format": message_format,
            }

            # Include session_id if we have one
            if self.session_id:
                payload["session_id"] = self.session_id

            response = requests.post(
                f"{self.base_url}/api/v1/a2a/message",
                headers=self._get_headers(),
                json=payload,
                timeout=30
            )

            if response.status_code == 200:
                message_response = response.json()
                print("   ✅ Success!")
                print(f"   Message ID: {message_response.get('message_id')}")
                print(f"   Response Format: {message_response.get('format')}")
                print(f"   Response Content:")
                print(f"   {message_response.get('content')[:200]}...")

                # Extract session_id for future messages
                if 'metadata' in message_response and 'session_id' in message_response['metadata']:
                    self.session_id = message_response['metadata']['session_id']
                    print(f"   Session ID: {self.session_id}")

                return message_response

            elif response.status_code == 401:
                print(f"   ❌ Authentication Failed (401)")
                print(f"   Response: {response.json()}")
                print(f"   💡 Hint: You may need to provide an API key with --api-key")
                return None

            else:
                print(f"   ❌ Failed: HTTP {response.status_code}")
                print(f"   Response: {response.text}")
                return None

        except Exception as e:
            print(f"   ❌ Error: {str(e)}")
            return None

    def validate_agent_card_compliance(self, agent_card: dict) -> bool:
        """
        Test T101: Validate agent card format compliance
        """
        print("\n🔍 Test 3: Validating Agent Card Compliance...")

        required_fields = ["id", "name", "description", "capabilities", "protocol_version"]
        missing_fields = []

        for field in required_fields:
            if field not in agent_card:
                missing_fields.append(field)

        if missing_fields:
            print(f"   ❌ Missing required fields: {', '.join(missing_fields)}")
            return False

        # Check capabilities
        if not isinstance(agent_card.get("capabilities"), list):
            print("   ❌ 'capabilities' must be a list")
            return False

        # Check metadata
        if "metadata" in agent_card:
            metadata = agent_card["metadata"]
            if "endpoints" not in metadata:
                print("   ⚠️  Warning: 'metadata.endpoints' not found")

        print("   ✅ Agent card format is compliant!")
        return True

    def check_health(self) -> bool:
        """
        Check A2A service health
        """
        print("\n🏥 Health Check...")
        print(f"   URL: {self.base_url}/api/v1/a2a/health")

        try:
            response = requests.get(
                f"{self.base_url}/api/v1/a2a/health",
                timeout=10
            )

            if response.status_code == 200:
                health = response.json()
                print("   ✅ Service is healthy")
                print(f"   Status: {health.get('status')}")
                print(f"   Protocol: {health.get('protocol')}")
                print(f"   Version: {health.get('version')}")
                print(f"   Active Sessions: {health.get('active_sessions')}")
                return True
            else:
                print(f"   ❌ Health check failed: HTTP {response.status_code}")
                return False

        except Exception as e:
            print(f"   ❌ Error: {str(e)}")
            return False


def run_tests(base_url: str, api_key: Optional[str] = None):
    """
    Run all A2A protocol tests
    """
    print("=" * 60)
    print("🚀 A2A Protocol Client Test Suite")
    print("=" * 60)
    print(f"Base URL: {base_url}")
    print(f"Authentication: {'Enabled (API Key provided)' if api_key else 'Disabled'}")
    print("=" * 60)

    client = A2AClient(base_url, api_key)

    # Test 0: Health check
    if not client.check_health():
        print("\n❌ Server is not available. Please start the server first.")
        print("   Run: cd backend && python -m src.server.app")
        return False

    # Test 1: Get agent card (T099)
    agent_card = client.get_agent_card()
    if not agent_card:
        print("\n❌ Failed to get agent card")
        return False

    # Test 3: Validate agent card (T101)
    if not client.validate_agent_card_compliance(agent_card):
        return False

    # Test 2: Send messages (T100)
    test_messages = [
        ("Hello! How are you?", "plain"),
        ("What can you do?", "plain"),
        ("Show me a code example", "plain"),
    ]

    for content, fmt in test_messages:
        response = client.send_message(content, fmt)
        if not response:
            print(f"\n❌ Failed to send message: {content}")
            return False

    print("\n" + "=" * 60)
    print("✅ All tests passed successfully!")
    print("=" * 60)
    print("\n📊 Summary:")
    print("   • Agent card retrieved and validated")
    print("   • Messages sent and received")
    print(f"   • Session maintained: {client.session_id}")
    print("   • A2A protocol compliance verified")

    return True


def main():
    parser = argparse.ArgumentParser(
        description="Test A2A protocol endpoints"
    )
    parser.add_argument(
        "--base-url",
        default="http://localhost:8000",
        help="Base URL of the A2A server (default: http://localhost:8000)"
    )
    parser.add_argument(
        "--api-key",
        help="API key for authentication (optional)"
    )
    parser.add_argument(
        "--verbose",
        action="store_true",
        help="Enable verbose output"
    )

    args = parser.parse_args()

    try:
        success = run_tests(args.base_url, args.api_key)
        sys.exit(0 if success else 1)
    except KeyboardInterrupt:
        print("\n\n⚠️  Tests interrupted by user")
        sys.exit(1)
    except Exception as e:
        print(f"\n\n❌ Unexpected error: {str(e)}")
        import traceback
        if args.verbose:
            traceback.print_exc()
        sys.exit(1)


if __name__ == "__main__":
    main()
